ZEP-XXX-journal-record-format.md

```diff
@@ -0,0 +1,239 @@
+```
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+| Header |  sbe_headers of metadata | checksum | length | sbe_headers of Data | index | asqn | length | sbe_headers of RaftRecord | term | entrytype | sbe_headers of ConfigurationEntry | timestamp | members |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+```
+
+A record with an InitialEntry
+```
++-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+| Header |  sbe_headers of metadata | checksum | length | sbe_headers of Data | index | asqn | length | sbe_headers of RaftRecord | term | entrytype |
++-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+```
+
+## Replication schema
+
+One approach is to replicate the serialized journal record as it is. However the raft module which sends and receives the replicated record needs to have access to raft term.
+Hence, the record needs to be (partially) deserialized at the raft layer.
+The serialization logic should be then shared by the raft and the journal.
+Currently the transport layer uses Kryo for serialization.
+So even if we replicate the journal record as it is, we are not getting much benefits from it.
+We would still need to copy the record before it is send.
+However, if we can make use of operating systems' feature to transfer bytes directly from the file to the network it would make sense to transfer the entire journal record as it is.
+But currently we don't see any benefit in doing so.
+Hence we propose to have a separate schema for replication.
+
+A replicated record contains following fields:
+1. term (long)
+2. index (long)
+3. asqn (asqn)
+4. checksum (long)
+5. serializedRaftRecord (byte[])
+
+Here `serializedRaftRecord` is the sbe serialized bytes of RaftRecord with the following schema.
+```
++ --------------------------------------------------
+| term<uint64> | entry type | entry<SBE(RaftEntry)> |
++ --------------------------------------------------
+```
+
+The raft thread that receives this record uses the term and index to verify the preconditions before writing it to the journal. It should then construct a journal record using `index`, `asqn`, `checksum` and `serializedRaftRecord` which can be appended to the journal. Journal is expected to verify if the index and checksum is correct.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
We might want to add a section about alternatives in terms of serialization format - why zero copy? Why SBE and not Flatbuffers or Cap'n'Proto?

I can do that as I already had a look at these.

Reply:
Thanks. That will be good.